### PR TITLE
Canbus base

### PR DESCRIPTION
**Pull request by mvturnho**
## Description:
Implementation to interconnect esphome devices using canbus.
This is the base platform component implementation for the canbus platform component.
Without the physical canbus (mcp2515) component it does not work.
So to start testing this "canbus_base" you also need the canbus_mcp2515 component implementation .

you may test this code by checking out the full canbus that includes the mcp2515 implementation. 
All code changes are first tested in canbus and then commited in canbus_base or canbus_mcp2515

**Related issue (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Checklist:
  - [ ] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by mvturnho** [comment]
@OttoWinter If I can be of any additional service to get this commited please let me know.

File: esphome/components/canbus/__init__.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
+import re
```
Comment (OttoWinter) [inline esphome/components/canbus/__init__.py]: https://travis-ci.org/esphome/esphome/jobs/634403707#L378

File: esphome/components/canbus/__init__.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+import esphome.codegen as cg
+import esphome.config_validation as cv
+from esphome import automation
+from esphome.automation import maybe_simple_id
```
Comment (OttoWinter) [inline esphome/components/canbus/__init__.py]: https://travis-ci.org/esphome/esphome/jobs/634403707#L379

File: esphome/components/canbus/__init__.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        cg.add(var.set_data_template(templ))
+    else:
+        cg.add(var.set_data_static(data))
+    yield var
+
+
+@coroutine_with_priority(100.0)
+def to_code(config):
+    cg.add_global(canbus_ns.using)
+    cg.add_define("USE_CANBUS")
```
Comment (OttoWinter) [inline esphome/components/canbus/__init__.py]: https://travis-ci.org/esphome/esphome/jobs/634403707#L380

Add a newline at the end of the file

File: esphome/components/canbus/__init__.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
+import re
+
+import esphome.codegen as cg
+import esphome.config_validation as cv
+from esphome import automation
+from esphome.automation import maybe_simple_id
+from esphome.core import CORE, coroutine, coroutine_with_priority
+from esphome.py_compat import text_type, binary_type, char_to_byte
```
Comment (OttoWinter) [inline esphome/components/canbus/__init__.py]: https://travis-ci.org/esphome/esphome/jobs/634403707#L386

These types have been removed when Python 2 support was dropped.

`text_type` is now `str`
`binary_type` is now `bytes`
`char_to_byte` is now `ord`

File: esphome/components/canbus/canbus.h [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+class Canbus : public Component {
+public:
+  Canbus(){};
+  void setup() override;
+  void dump_config() override;
+  float get_setup_priority() const override { return setup_priority::HARDWARE; }
+  void loop() override;
+
+  void send_data(uint32_t can_id, const std::vector<uint8_t> data);
```
Comment (OttoWinter) [inline esphome/components/canbus/canbus.h]:
```suggestion
  void send_data(uint32_t can_id, const std::vector<uint8_t> &data);
```

https://travis-ci.org/esphome/esphome/jobs/634403709#L868

File: esphome/components/canbus/canbus.h [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+  void set_bitrate(CanSpeed bit_rate) { this->bit_rate_ = bit_rate; }
+
+  void add_trigger(CanbusTrigger *trigger);
+
+protected:
+  std::vector<CanbusTrigger *> triggers_{};
+  uint32_t sender_id_{0};
+  CanSpeed bit_rate_{CAN_125KBPS};
+
+  virtual bool setup_internal_();
```
Comment (OttoWinter) [inline esphome/components/canbus/canbus.h]:
```suggestion
  bool setup_internal_() override;
```

Equivalent, but preferred style.

File: esphome/components/canbus/canbus.cpp [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+void Canbus::setup() {
+  ESP_LOGCONFIG(TAG, "Setting up Canbus...");
+  if (!this->setup_internal_()) {
+    ESP_LOGE(TAG, "Canbus setup error!");
+    this->mark_failed();
+  }
+}
+
+void Canbus::dump_config() {
```
Comment (OttoWinter) [inline esphome/components/canbus/canbus.cpp]: https://travis-ci.org/esphome/esphome/jobs/634403709#L934

Reformat needed to pass CI checks

**Review by OttoWinter — COMMENTED** [review]
Hi @mvturnho!

Thanks for re-submitting this PR, and I want to say sorry for not reviewing the old PR earlier. I should have done so sooner.

The code looks good from an architectural perspective. Only thing left to do is make the lint checks pass. I've copied the links of the travis-ci lint log so that you can fix them more easily.

Thanks!

**Comment by mvturnho** [comment]
> Hi @mvturnho!
> 
> Thanks for re-submitting this PR, and I want to say sorry for not reviewing the old PR earlier. I should have done so sooner.
> 
> The code looks good from an architectural perspective. Only thing left to do is make the lint checks pass. I've copied the links of the travis-ci lint log so that you can fix them more easily.
> 
> Thanks!

No problem, I can imagine that you are extremely busy at times.
I will fix the lint errors and your review feedback.

File: esphome/components/canbus/canbus.h (hunk 6 above)
Comment (mvturnho) [reply esphome/components/canbus/canbus.h]: This does not work out.
The setup_internal_ is for the baseclass Canbus. In the MCP2515 and later Esp32Canbus we override this virtual function.

**Comment by mvturnho** [comment]
@OttoWinter I saw in other components that virtual protected functions are without the trailing underscores. So I changed that for the canbus base and mcp2515 components

**Comment by mvturnho** [comment]
@brandond @OttoWinter it seems github will be closing this PR again. Would it be possible to commit   the canbus functionality?

**Comment by glmnet** [comment]
docs?

**Comment by glmnet** [comment]
Closing as this is included in https://github.com/esphome/esphome/pull/968

**Pull request closed without merging**